File: ts/indicators/mark-jurik/jurik-directional-movement-index/jurik-directional-movement-index.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { JurikMovingAverage } from '../jurik-moving-average/jurik-moving-average';
import { JurikDirectionalMovementIndexParams } from './params';

/**
 * Jurik Directional Movement Index (DMX).
 *
 * Produces three output lines:
 *   - Bipolar: 100*(Plus-Minus)/(Plus+Minus)
 *   - Plus: JMA(upward) / JMA(TrueRange)
 *   - Minus: JMA(downward) / JMA(TrueRange)
 *
 * The internal JMA instances use phase=-100 (maximum lag, no overshoot).
 */
export class JurikDirectionalMovementIndex implements Indicator {
    private readonly _mnemonic: string;
    private readonly _description: string;
    private _primed = false;
    private bar = 0;
    private prevHigh = NaN;
    private prevLow = NaN;
    private prevClose = NaN;
    private readonly jmaPlus: JurikMovingAverage;
    private readonly jmaMinus: JurikMovingAverage;
    private readonly jmaDenom: JurikMovingAverage;
    private plusVal = NaN;
    private minusVal = NaN;
    private bipolarVal = NaN;

    constructor(params: JurikDirectionalMovementIndexParams) {
        const length = params.length;

        if (length < 1) {
            throw new Error('invalid jurik directional movement index parameters: length should be positive');
        }

        this._mnemonic = `dmx(${length})`;
        this._description = 'Jurik directional movement index ' + this._mnemonic;

        const jmaParams = { length, phase: -100 };
        this.jmaPlus = new JurikMovingAverage(jmaParams);
        this.jmaMinus = new JurikMovingAverage(jmaParams);
        this.jmaDenom = new JurikMovingAverage(jmaParams);
    }

    /** The indicator mnemonic. */
    public get mnemonic(): string {
        return this._mnemonic;
    }

    /** The indicator description. */
    public get description(): string {
        return this._description;
    }

    /** Indicates whether the indicator is primed. */
    public isPrimed(): boolean {
        return this._primed;
    }

    /** Describes the output data of the indicator. */
    public metadata(): IndicatorMetadata {
        return buildMetadata(
            IndicatorIdentifier.JurikDirectionalMovementIndex,
            this._mnemonic,
            this._description,
            [
                { mnemonic: this._mnemonic + ':bipolar', description: this._description + ' bipolar' },
                { mnemonic: this._mnemonic + ':plus', description: this._description + ' plus' },
                { mnemonic: this._mnemonic + ':minus', description: this._description + ' minus' },
            ],
        );
    }

    /** Updates the indicator given the next high, low, and close values. */
    public update(high: number, low: number, close: number): [number, number, number] {
        const warmup = 41;
        const epsilon = 0.00001;
        const hundred = 100.0;

        this.bar++;

        let trueRange = 0;
        let upward = 0;
        let downward = 0;

        if (this.bar >= 2) {
            const v1 = hundred * (high - this.prevHigh);
            const v2 = hundred * (this.prevLow - low);

            if (v1 > v2 && v1 > 0) {
                upward = v1;
            }

            if (v2 > v1 && v2 > 0) {
                downward = v2;
            }
        }

        if (this.bar >= 3) {
            const m1 = Math.abs(high - low);
            const m2 = Math.abs(high - this.prevClose);
            const m3 = Math.abs(low - this.prevClose);
            trueRange = Math.max(Math.max(m1, m2), m3);
        }

        this.prevHigh = high;
        this.prevLow = low;
        this.prevClose = close;

        // Feed into JMA instances.
        const numerPlus = this.jmaPlus.update(upward);
        const numerMinus = this.jmaMinus.update(downward);
        const denom = this.jmaDenom.update(trueRange);

        if (this.bar <= warmup) {
            this.bipolarVal = NaN;
            this.plusVal = NaN;
            this.minusVal = NaN;

            return [NaN, NaN, NaN];
        }

        this._primed = true;

        // Compute Plus and Minus.
        if (denom > epsilon) {
            this.plusVal = numerPlus / denom;
        } else {
            this.plusVal = 0;
        }

        if (denom > epsilon) {
            this.minusVal = numerMinus / denom;
        } else {
            this.minusVal = 0;
        }

        // Compute Bipolar.
        const sum = this.plusVal + this.minusVal;
        if (sum > epsilon) {
            this.bipolarVal = hundred * (this.plusVal - this.minusVal) / sum;
        } else {
            this.bipolarVal = 0;
        }

        return [this.bipolarVal, this.plusVal, this.minusVal];
    }

    /** Updates the indicator given the next scalar sample. */
    public updateScalar(sample: Scalar): IndicatorOutput {
        const v = sample.value;
        const [bipolar, plus, minus] = this.update(v, v, v);

        const s1 = new Scalar();
        s1.time = sample.time;
        s1.value = bipolar;
        const s2 = new Scalar();
        s2.time = sample.time;
        s2.value = plus;
        const s3 = new Scalar();
        s3.time = sample.time;
        s3.value = minus;

        return [s1, s2, s3];
    }

    /** Updates the indicator given the next bar sample. */
    public updateBar(sample: Bar): IndicatorOutput {
        const [bipolar, plus, minus] = this.update(sample.high, sample.low, sample.close);

        const s1 = new Scalar();
        s1.time = sample.time;
        s1.value = bipolar;
        const s2 = new Scalar();
        s2.time = sample.time;
        s2.value = plus;
        const s3 = new Scalar();
        s3.time = sample.time;
        s3.value = minus;

        return [s1, s2, s3];
    }

    /** Updates the indicator given the next quote sample. */
    public updateQuote(sample: Quote): IndicatorOutput {
        const [bipolar, plus, minus] = this.update(sample.askPrice, sample.bidPrice, (sample.askPrice + sample.bidPrice) / 2);

        const s1 = new Scalar();
        s1.time = sample.time;
        s1.value = bipolar;
        const s2 = new Scalar();
        s2.time = sample.time;
        s2.value = plus;
        const s3 = new Scalar();
        s3.time = sample.time;
        s3.value = minus;

        return [s1, s2, s3];
    }

    /** Updates the indicator given the next trade sample. */
    public updateTrade(sample: Trade): IndicatorOutput {
        const v = sample.price;
        const [bipolar, plus, minus] = this.update(v, v, v);

        const s1 = new Scalar();
        s1.time = sample.time;
        s1.value = bipolar;
        const s2 = new Scalar();
        s2.time = sample.time;
        s2.value = plus;
        const s3 = new Scalar();
        s3.time = sample.time;
        s3.value = minus;

        return [s1, s2, s3];
    }
}

File: go/indicators/markjurik/jurikdirectionalmovementindex/jurikdirectionalmovementindex.go
package jurikdirectionalmovementindex

import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/markjurik/jurikmovingaverage"
)

// JurikDirectionalMovementIndex computes the Jurik directional movement index (DMX),
// see http://jurikres.com/.
//
// It produces three output lines:
//   - Bipolar: 100*(Plus-Minus)/(Plus+Minus)
//   - Plus: JMA(upward) / JMA(TrueRange)
//   - Minus: JMA(downward) / JMA(TrueRange)
//
// The internal JMA instances use phase=-100 (maximum lag, no overshoot).
type JurikDirectionalMovementIndex struct {
	mu          sync.RWMutex
	mnemonic    string
	description string
	primed      bool
	bar         int
	prevHigh    float64
	prevLow     float64
	prevClose   float64
	jmaPlus     *jurikmovingaverage.JurikMovingAverage
	jmaMinus    *jurikmovingaverage.JurikMovingAverage
	jmaDenom    *jurikmovingaverage.JurikMovingAverage
	plusVal     float64
	minusVal    float64
	bipolarVal  float64
}

// NewJurikDirectionalMovementIndex returns an instance of the indicator created using supplied parameters.
func NewJurikDirectionalMovementIndex(p *JurikDirectionalMovementIndexParams) (*JurikDirectionalMovementIndex, error) {
	return newJurikDirectionalMovementIndex(p.Length)
}

func newJurikDirectionalMovementIndex(length int) (*JurikDirectionalMovementIndex, error) {
	const (
		invalid = "invalid jurik directional movement index parameters"
		fmts    = "%s: %s"
fmtn = "jdmx(%d)"
		phase   = -100
		minlen  = 1
	)

	if length < minlen {
		return nil, fmt.Errorf(fmts, invalid, "length should be positive")
	}

	jmaParams := &jurikmovingaverage.JurikMovingAverageParams{
		Length: length,
		Phase:  phase,
	}

	jmaPlus, err := jurikmovingaverage.NewJurikMovingAverage(jmaParams)
	if err != nil {
		return nil, fmt.Errorf(fmts, invalid, err.Error())
	}

	jmaMinus, err := jurikmovingaverage.NewJurikMovingAverage(jmaParams)
	if err != nil {
		return nil, fmt.Errorf(fmts, invalid, err.Error())
	}

	jmaDenom, err := jurikmovingaverage.NewJurikMovingAverage(jmaParams)
	if err != nil {
		return nil, fmt.Errorf(fmts, invalid, err.Error())
	}

	mnemonic := fmt.Sprintf(fmtn, length)

	return &JurikDirectionalMovementIndex{
		mnemonic:    mnemonic,
		description: "Jurik directional movement index " + mnemonic,
		jmaPlus:     jmaPlus,
		jmaMinus:    jmaMinus,
		jmaDenom:    jmaDenom,
		prevHigh:    math.NaN(),
		prevLow:     math.NaN(),
		prevClose:   math.NaN(),
	}, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *JurikDirectionalMovementIndex) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *JurikDirectionalMovementIndex) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.JurikDirectionalMovementIndex,
		s.mnemonic,
		s.description,
		[]core.OutputText{
			{Mnemonic: s.mnemonic + ":bipolar", Description: s.description + " bipolar"},
			{Mnemonic: s.mnemonic + ":plus", Description: s.description + " plus"},
			{Mnemonic: s.mnemonic + ":minus", Description: s.description + " minus"},
		},
	)
}

// Update updates the indicator given the next high, low, and close values.
func (s *JurikDirectionalMovementIndex) Update(high, low, close float64) (bipolar, plus, minus float64) { //nolint:cyclop
	s.mu.Lock()
	defer s.mu.Unlock()

	const (
		warmup  = 41
		epsilon = 0.00001
		hundred = 100.0
	)

	s.bar++

	// TrueRange starts from bar 2 (needs prevClose from bar before previous).
	// Upward/downward starts from bar 1 (needs prevHigh/prevLow).
	var trueRange, upward, downward float64

	if s.bar >= 2 { //nolint:mnd
		// Upward/downward movement.
		v1 := hundred * (high - s.prevHigh)
		v2 := hundred * (s.prevLow - low)

		if v1 > v2 && v1 > 0 {
			upward = v1
		}

		if v2 > v1 && v2 > 0 {
			downward = v2
		}
	}

	if s.bar >= 3 { //nolint:mnd
		// True range (starts from bar 2 in reference — needs 2 previous bars).
		m1 := math.Abs(high - low)
		m2 := math.Abs(high - s.prevClose)
		m3 := math.Abs(low - s.prevClose)
		trueRange = math.Max(math.Max(m1, m2), m3)
	}

	s.prevHigh = high
	s.prevLow = low
	s.prevClose = close

	// Feed into JMA instances.
	numerPlus := s.jmaPlus.Update(upward)
	numerMinus := s.jmaMinus.Update(downward)
	denom := s.jmaDenom.Update(trueRange)

	if s.bar <= warmup {
		s.bipolarVal = math.NaN()
		s.plusVal = math.NaN()
		s.minusVal = math.NaN()

		return math.NaN(), math.NaN(), math.NaN()
	}

	s.primed = true

	// Compute Plus and Minus.
	if denom > epsilon {
		s.plusVal = numerPlus / denom
	} else {
		s.plusVal = 0
	}

	if denom > epsilon {
		s.minusVal = numerMinus / denom
	} else {
		s.minusVal = 0
	}

	// Compute Bipolar.
	sum := s.plusVal + s.minusVal
	if sum > epsilon {
		s.bipolarVal = hundred * (s.plusVal - s.minusVal) / sum
	} else {
		s.bipolarVal = 0
	}

	return s.bipolarVal, s.plusVal, s.minusVal
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *JurikDirectionalMovementIndex) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value

	return s.updateEntity(sample.Time, v, v, v)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *JurikDirectionalMovementIndex) UpdateBar(sample *entities.Bar) core.Output {
	return s.updateEntity(sample.Time, sample.High, sample.Low, sample.Close)
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *JurikDirectionalMovementIndex) UpdateQuote(sample *entities.Quote) core.Output {
	return s.updateEntity(sample.Time, sample.Ask, sample.Bid, (sample.Ask+sample.Bid)/2) //nolint:mnd
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *JurikDirectionalMovementIndex) UpdateTrade(sample *entities.Trade) core.Output {
	v := sample.Price

	return s.updateEntity(sample.Time, v, v, v)
}

func (s *JurikDirectionalMovementIndex) updateEntity(
	t time.Time, high, low, close float64,
) core.Output {
	const outputCount = 3

	output := make([]any, outputCount)
	bipolar, plus, minus := s.Update(high, low, close)

	i := 0
	output[i] = entities.Scalar{Time: t, Value: bipolar}
	i++
	output[i] = entities.Scalar{Time: t, Value: plus}
	i++
	output[i] = entities.Scalar{Time: t, Value: minus}

	return output
}
